### PR TITLE
Fix type conversion during communication for regular get, send, sendget.

### LINKED ISSUE #292: gfortran 7 regression: no type conversion before coarray put
This issue is for tracking gfortran [bug report 78892], which was first reported in an OpenCoarrays [mailing list message].  The gfortran bug report contains the following text:
```
The code below demonstrates a regression in which gfortran 7.0.0 20161215 is not performing a necessary implicit type conversion before putting data on a remote image.  By contrast, the code below works as expected with gfortran 5.4.0 and 6.2.0. 

$ cat convert-before-put.f90 
  real :: a[*]
  integer :: receiver
  associate(me=>this_image())
    if (me == 1) then
      do receiver = 2, num_images()
        a[receiver] = receiver ! implicit real(receiver) needed here
        sync images (receiver) ! notify remote image that data has been put
      end do 
    else
      sync images (1) ! await notification of data put by image 1
      if (a/=real(me)) print *, "Image ",me,": received ",a,", but expected ",real(me)
    end if
  end associate
end 

$ caf convert-before-put.f90 

$ cafrun -np 2 ./a.out
 Image            2 : received    0.00000000     , but expected    2.00000000
```
[mailing list message]: https://groups.google.com/forum/#!topic/opencoarrays/6E1RLQZqkE4
[bug report 78892]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78892

### PR DESCRIPTION
**Pull request by vehre**
[links]:#
[contributing guidelines]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/CONTRIBUTING.md
[issue]: https://github.com/sourceryinstitute/OpenCoarrays/issues
[PR response img]: https://img.shields.io/issuestats/p/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square
[coverage]: https://img.shields.io/codecov/c/github/sourceryinstitute/OpenCoarrays/master.svg?style=flat-square

|  Avg response time                |  coverage on master         |
|:---------------------------------:|:---------------------------:|
| ![Issue Stats][PR response img]   | ![Codecov branch][coverage] |

## Summary of changes ##

Fix #292: Missing type/kind conversion in communication primitives `get()`, `send()`, `sendget()`.
Added copy-to-self support without call of MPI-library-functions.
Added char-array support.
Added full vector-index support for the three primitives.
Added optimized strided communication support where ever possible with fall back to regular communication pattern.
Fix (maybe? needs testing) #427: Remote indirect addressing

## Rationale for changes ##

Support was missing or crappy.

## Additional info and certifications ##

This pull request (PR) is a:

- [x] Bug fix
- [x] Feature addition
- [ ] Other, Please describe:

### I certify that ###

- [x] I reviewed and followed the [contributing guidelines], including
      - Increasing test coverage for all feature-addition PRs
      - Increasing test coverage for all bug-fix PRs for which there
        does not already exist a related test that failed before the PR
      - At least maintaining test coverage for all other PRs
      - Ensuring that all tests pass when run locally 
      - Naming PR  to indicate work in progress (WIP) and to attach the PR
        to the appropriate bug report or feature request [issue]
      - White space (no trailing white space or white space errors may
        be introduced)
      - Commenting code where it is non-obvious and non-trivial
      - Logically atomic, self consistent and coherent commits
      - Commit message content
      - Waiting 24 hours before self-approving the PR to give another
        OpenCoarrays developer a chance to review my proposed code

### CODE REVIEW

**Comment by vehre** [comment]
Note, for Travis to complete successfully, a rather new gcc-7 is needed: gcc >= 7.3.

File: src/libcaf-gfortran-descriptor.h [hunk 1]
```diff
@@ -51,6 +51,9 @@ typedef struct gfc_descriptor_t {
   void *base_addr;
   size_t offset;
   ptrdiff_t dtype;
+#if (__GNUC__ >= 8)
```
Comment (rouson) [inline src/libcaf-gfortran-descriptor.h]: @vehre Is there anything that would be useful for a Fortran programmer to know about the behavior of OpenCoarrays with  GCC >= 8?  If so, let's document the differences somewhere.  If it's in a commit message, then it's likely to make it into the next release notes.  If not, then I'm not sure of the best place but probably either README.md or INSTALL.md.

File: src/libcaf-gfortran-descriptor.h [hunk 2]
```diff
@@ -51,7 +51,7 @@ typedef struct gfc_descriptor_t {
   void *base_addr;
   size_t offset;
   ptrdiff_t dtype;
-#if (__GNUC__ >= 8)
+#ifdef GCC_GE_8
   ptrdiff_t span;
```
Comment (rouson) [inline src/libcaf-gfortran-descriptor.h]: I think this answers my earlier question.  Apparently, the only change with GCC>=8 is internal: a new descriptor member `span` (added during Paul's work on PDT support if I recall correctly).  I don't imagine therein any reason to be mention this change in user documentation.

**Comment by rouson** [comment]
@zbeekman I think there's a [shell command error] in the Travis scripts.

[shell command error]: https://travis-ci.org/sourceryinstitute/OpenCoarrays/jobs/307514477#L489

File: src/libcaf-gfortran-descriptor.h (hunk 2 above)
Comment (vehre) [reply src/libcaf-gfortran-descriptor.h]: Well, none that I know of. Most of the development in gcc-8 went past me. Besides two small bugfixes there is nothing that jumps into my mind.

**Comment by vehre** [comment]
No, that error is unfortunately triggered by one of the two fixes I applied to gfortran missing. One needs to upgrade the compiler on travis, or deactivate the test, when being compiled with gcc-7 < 7.3.

**Comment by zbeekman** [comment]
@vehre Even with 7.3 one of the tests (character kind conversion) is failing: https://travis-ci.org/sourceryinstitute/OpenCoarrays/jobs/307514478#L1071

I'll work on ensuring that the other tests are only run under the correct circumstances.

**Comment by rouson** [comment]
I worked on cordoning off the tests yesterday, but didn’t quite get to the point of pushing a commit so we can look at what I did when we talk later today if you like.

**Comment by vehre** [comment]
That's odd, because the fix was in time for 7.3.

**Comment by zbeekman** [comment]
> That's odd, because the fix was in time for 7.3.

Oops, I lied. It's still 7.2. I'll look into whether or not a PR for 7.3 is open with Homebrew yet.

**Comment by zbeekman** [comment]
I'm going to work on updating this branch, and ensuring that the tests are only run under the right conditions.

File: CMakeLists.txt [hunk 3]
```diff
@@ -673,7 +690,7 @@ if(opencoarrays_aware_compiler)
   # Integration tests verifying the use of libcaf_mpi in applications
   add_caf_test(hello_multiverse 2 hello_multiverse)
   add_caf_test(coarray_burgers_pde 2 coarray_burgers_pde)
-  add_caf_test(co_heat 2 co_heat)
+#  add_caf_test(co_heat 2 co_heat)
```
Comment (zbeekman) [inline CMakeLists.txt]: Why was this test (co_heat) disabled?

Comment (vehre) [reply CMakeLists.txt]: Er, I don't know exactly. I am in the impression, that it was done by a merge of master?

Comment (zbeekman) [reply CMakeLists.txt]: @vehre No, you did it in https://github.com/sourceryinstitute/OpenCoarrays/commit/7197bc7c086fe30ebd7079f3ac30d3389bda9de1

Comment (zbeekman) [reply CMakeLists.txt]: I'll reenable it if there was no intentional reason.

File: src/tests/unit/send-get/CMakeLists.txt [hunk 4]
```diff
@@ -3,11 +3,35 @@ caf_compile_executable(get_array get_array_test.f90)
...
+caf_compile_executable(get_with_vector_index get_with_vector_index.f90)
+
+## Pure send() tests
+caf_compile_executable(send_array send_array_test.f90)
+caf_compile_executable(send_convert_char_array send_convert_char_array.f90)
+caf_compile_executable(send_with_vector_index send_with_vector_index.f90)
+
+# Pure sendget() tests
+caf_compile_executable(sendget_convert_char_array sendget_convert_char_array.f90)
+#caf_compile_executable(sendget_with_vector_index sendget_with_vector_index.f90)
```
Comment (zbeekman) [inline src/tests/unit/send-get/CMakeLists.txt]: @vehre Is there an additional test for this that wasn't committed? (sendget_with_vector_index.f90)

**Comment by vehre** [comment]
Why does the tests start with test no 21 and not with no 1?
No, I haven't tested against all gcc-versions possibly available.
But the major issue is fixed with commit 7158ea2.

File: src/tests/unit/send-get/CMakeLists.txt (hunk 4 above)
Comment (vehre) [reply src/tests/unit/send-get/CMakeLists.txt]: Nope, none that I can find. Presumably I just planned to do it, but never did and the commented cmake-line just remained there out of laziness.

File: CMakeLists.txt (hunk 3 above)
Comment (vehre) [reply CMakeLists.txt]: May be it was just waking up my testing and I just forgot to remove the # on commit.

**Comment by zbeekman** [comment]
> Why does the tests start with test no 21 and not with no 1?

That is a very good question. I have no idea. Travis can "eat" output sometimes. If this persists I will look into it. Tests can be parallelized, but, as far as I can tell, we have not asked CTest to run them in parallel so the output should be in ascending numeric order.

> No, I haven't tested against all gcc-versions possibly available.

We don't need all versions just latest GCC 6 and 7. 6 and 7 is what the CI tests test. Those are the versions we currently need to support, at least until major regressions associated with 7 are resolved.

> But the major issue is fixed with commit 7158ea2.

Awesome, thanks!

**Comment by vehre** [comment]
send(get)_convert_char_array is failing on gcc-6 because the compiler does not pass the char array as expected. So those tests will only pass on gcc >= 7.

**Comment by zbeekman** [comment]
> send(get)_convert_char_array is failing on gcc-6 because the compiler does not pass the char array as expected. So those tests will only pass on gcc >= 7.
> 
> Merge state

Ah, OK, I'll add the appropriate guards so that those tests don't get run for GFortran < 7.3 Thanks

**Comment by zbeekman** [comment]
@vehre Which tests are supposed to *REQUIRE* >= 7.3? Right now, two tests fail with GCC 7.2:

 - `send_convert_char_array`
 - `send_with_vector_index`

Does this sound right?

The character conversion tests ICE GCC 6.x but I guess that is to be expected.

**Comment by zbeekman** [comment]
@vehre if you can confirm that this sounds right, I'll approve and merge.
> @vehre Which tests are supposed to REQUIRE >= 7.3? Right now, two tests fail with GCC 7.2:
> 
> send_convert_char_array
> send_with_vector_index
> Does this sound right?
> 
> The character conversion tests ICE GCC 6.x but I guess that is to be expected.

**Comment by vehre** [comment]
Well, I am wondering a bit that send_with_vector_index is failing. That is odd. I am at work at the moment, so I can't investigate further at the moment.

**Comment by zbeekman** [comment]
To clarify, this is not being tested with GFortran 7.3 yet. Also, some
commits have been pushed to your branch by Damian and me. If you run into
issues merging when you pull I'm happy to help out.

Thanks!

**Comment by vehre** [comment]
send_with_vector_index.f90 is not supposed to fail with gcc >= 6. A sync all was missing here.

**Pull request merged**